File: Data Structures & Algorithms/count-paths/submission-6.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        memo = [[-1 for c in range(n)] for r in range(m)]
        def dfs(r, c):
            if r == m - 1 or c == n - 1: return 1
            elif memo[r][c] != -1: return memo[r][c]
            else:
                memo[r][c] = dfs(r+1, c) + dfs(r, c+1)
                return memo[r][c]

        return dfs(0, 0)

File: Data Structures & Algorithms/combination-target-sum/submission-6.java
class Solution {
    private List<List<Integer>> ret;
    private List<Integer> currCombination;
    private int currSum;

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        ret = new ArrayList<>();
        currCombination = new ArrayList<>();
        currSum = 0;

        dfs(nums, target, 0);

        return ret;
    }

    private void dfs(int[] nums, int target, int i) {
        if (currSum == target) ret.add(new ArrayList<Integer>(currCombination));
        else if (currSum > target || i == nums.length) return;
        else {
            currCombination.add(nums[i]);
            currSum += nums[i];
            dfs(nums, target, i);

            currCombination.remove(currCombination.size() - 1);
            currSum -= nums[i];
            dfs(nums, target, i+1);
        }
    }
}
